File: mobile/lib/core/models/api_response.dart
class ApiError {
  final String code;
  final String message;

  ApiError({required this.code, required this.message});

  factory ApiError.fromJson(Map<String, dynamic> json) =>
      ApiError(code: json['code'] as String, message: json['message'] as String);
}

class PageMeta {
  final int total;
  final int page;
  final int limit;
  final int totalPages;

  PageMeta({required this.total, required this.page, required this.limit, required this.totalPages});

  factory PageMeta.fromJson(Map<String, dynamic> json) => PageMeta(
        total: json['total'] as int,
        page: json['page'] as int,
        limit: json['limit'] as int,
        totalPages: json['totalPages'] as int,
      );
}

class ApiResponse<T> {
  final bool success;
  final T? data;
  final PageMeta? meta;
  final ApiError? error;

  ApiResponse({required this.success, this.data, this.meta, this.error});

  factory ApiResponse.fromJson(Map<String, dynamic> json, T Function(dynamic) fromJsonT) {
    return ApiResponse(
      success: json['success'] as bool,
      data: json['data'] == null ? null : fromJsonT(json['data']),
      meta: json['meta'] == null ? null : PageMeta.fromJson(json['meta'] as Map<String, dynamic>),
      error: json['error'] == null ? null : ApiError.fromJson(json['error'] as Map<String, dynamic>),
    );
  }
}

File: backend/FitBodyApp.Application/Common/ApiResponse.cs
namespace FitBodyApp.Application.Common;

public class ApiError
{
    public string Code { get; set; } = string.Empty;
    public string Message { get; set; } = string.Empty;
}

public class ApiResponse<T>
{
    public bool Success { get; set; }
    public T? Data { get; set; }
    public object? Meta { get; set; }
    public ApiError? Error { get; set; }

    public static ApiResponse<T> Ok(T data, object? meta = null) => new()
    {
        Success = true,
        Data = data,
        Meta = meta
    };

    public static ApiResponse<T> Fail(string code, string message) => new()
    {
        Success = false,
        Error = new ApiError { Code = code, Message = message }
    };
}

public class PageMeta
{
    public int Total { get; set; }
    public int Page { get; set; }
    public int Limit { get; set; }
    public int TotalPages { get; set; }
}
